src/m365/spo/commands/listitem/listitem-roleassignment-add.ts

```diff
@@ -0,0 +1,279 @@
+        }
+
+        if (args.options.roleDefinitionId && isNaN(args.options.roleDefinitionId)) {
+          return `Specified roleDefinitionId ${args.options.roleDefinitionId} is not a number`;
+        }
+
+        const listOptions: any[] = [args.options.listId, args.options.listTitle, args.options.listUrl];
+        if (listOptions.some(item => item !== undefined) && listOptions.filter(item => item !== undefined).length > 1) {
+          return `Specify either list id or title or list url`;
+        }
+
+        if (listOptions.filter(item => item !== undefined).length === 0) {
+          return `Specify at least list id or title or list url`;
+        }
+
+        const principalOptions: any[] = [args.options.principalId, args.options.upn, args.options.groupName];
+        if (!principalOptions.some(item => item !== undefined)) {
+          return `Specify either principalId, upn or groupName`;
+        }
+
+        if (principalOptions.filter(item => item !== undefined).length > 1) {
+          return `Specify either principalId, upn or groupName but not multiple`;
+        }
+
+        const roleDefinitionOptions: any[] = [args.options.roleDefinitionId, args.options.roleDefinitionName];
+        if (!roleDefinitionOptions.some(item => item !== undefined)) {
+          return `Specify either roleDefinitionId or roleDefinitionName`;
+        }
+
+        if (roleDefinitionOptions.filter(item => item !== undefined).length > 1) {
+          return `Specify either roleDefinitionId or roleDefinitionName but not multiple`;
+        }
+
+        return true;
+      }
+    );
+  }
+
+  public async commandAction(logger: Logger, args: CommandArgs): Promise<void> {
+    if (this.verbose) {
+      logger.logToStderr(`Adding role assignment to listitem in site at ${args.options.webUrl}...`);
+    }
+
+    try {
+      let requestUrl: string = `${args.options.webUrl}/_api/web/`;
+      if (args.options.listId) {
+        requestUrl += `lists(guid'${formatting.encodeQueryParameter(args.options.listId)}')/`;
+      }
+      else if (args.options.listTitle) {
+        requestUrl += `lists/getByTitle('${formatting.encodeQueryParameter(args.options.listTitle)}')/`;
+      }
+      else if (args.options.listUrl) {
+        const listServerRelativeUrl: string = urlUtil.getServerRelativePath(args.options.webUrl, args.options.listUrl);
+        requestUrl += `GetList('${formatting.encodeQueryParameter(listServerRelativeUrl)}')/`;
+      }
+
+      requestUrl += `items(${args.options.listItemId})/`;
+
+      args.options.roleDefinitionId = await this.getRoleDefinitionId(args.options);
+      if (args.options.upn) {
+        args.options.principalId = await this.getUserPrincipalId(args.options);
+        await this.addRoleAssignment(requestUrl, args.options);
+      }
+      else if (args.options.groupName) {
+        args.options.principalId = await this.getGroupPrincipalId(args.options);
+        await this.addRoleAssignment(requestUrl, args.options);
```

Comment:
we agreed not to overwrite the args and keep them as read only. I need to correct this one